Add queryAll/getAll methods

alexkrolick opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**:

Adds "all" methods for:
- byAltText
- byTestId

<!-- Why are these changes necessary? -->

**Why**:

#27 

<!-- How were these changes implemented? -->

**How**:

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

* [x] Documentation
* [x] Tests
* [x] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
* [x] Added myself to contributors table <!-- this is optional, see the contributing guidelines for instructions -->

<!-- feel free to add additional comments -->

---

**Sidenote**

If these `getAll` methods are added, the `queryBy` methods seem kind of redundant - instead of checking if `queryBy` returns `null`, you can check if `getAll` is `[]`.

alexkrolick commented:
> I'm thinking that it would be good to actually have an *All* version for every query/get method for consistency.

Yep, just wanted to start with a few cases for review.

> And I'm thinking that we could probably have the *All* methods be the base and the existing methods just use those and return the first element to simplify the implementation.

Yeah. Would the `queryBy` methods continue to be supported (ie `if (result.length === 0) return null`) or phased out?

> Also, we'll want to make sure to document the existence of the *All* methods.

Yep.

kentcdodds commented:
Yes, let's do this with no breaking changes.

Yes, please maintain 100% code coverage 👍

alexkrolick commented:
OK, 🍏 

- Added `queryAll` methods for everything
- Reimplemented `query` APIs using `queryAll`
- Aliased `getAll` to `queryAll` (they are the same)
- Added tests for the array methods; most of the code is actually executed in tests for the `get` methods.

alexkrolick commented:
✅ Changed `getAll` to throw the same errors as the `get` methods

kentcdodds commented on src/queries.js:
```diff
@@ -10,156 +10,261 @@ function debugDOM(htmlElement) {
-        container,
-      )}`,
-    )
+    throw noMatchingAltText(container, alt)
```
What if instead of creating these error functions we had the `getBy*` functions use the `getAllBy*` functions? Perhaps using `firstResultOrNull`.

Sorry, I still haven't reviewed all the PR but this is just something else I noticed.

alexkrolick replied:
Yep that makes sense!

kentcdodds commented:
Thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the `other/MAINTAINING.md` and `CONTRIBUTING.md` files (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)

kentcdodds commented:
:tada: This PR is included in version 1.10.0 :tada:

The release is available on:
- `npm package (@latest dist-tag)`
- [GitHub release](https://github.com/kentcdodds/dom-testing-library/releases/tag/v1.10.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
